diff --git a/net/compat-drivers/drivers/net/wireless/ti/wl12xx/main.c b/net/compat-drivers/drivers/net/wireless/ti/wl12xx/main.c
index 2ba4e81..9d67143 100644
--- a/net/compat-drivers/drivers/net/wireless/ti/wl12xx/main.c
+++ b/net/compat-drivers/drivers/net/wireless/ti/wl12xx/main.c
@@ -599,9 +599,9 @@ static const int wl12xx_rtable[REG_TABLE_LEN] = {
 };
 
 /* TODO: maybe move to a new header file? */
-#define WL127X_FW_NAME_MULTI	"ti-connectivity/wl127x-fw-5-mr.bin"
-#define WL127X_FW_NAME_SINGLE	"ti-connectivity/wl127x-fw-5-sr.bin"
-#define WL127X_PLT_FW_NAME	"ti-connectivity/wl127x-fw-5-plt.bin"
+#define WL127X_FW_NAME_MULTI	"../../../firmware/ti-connectivity/wl127x-fw-5-mr.bin"
+#define WL127X_FW_NAME_SINGLE	"../../../firmware/ti-connectivity/wl127x-fw-5-sr.bin"
+#define WL127X_PLT_FW_NAME	"../../../firmware/ti-connectivity/wl127x-fw-5-plt.bin"
 
 #define WL128X_FW_NAME_MULTI	"ti-connectivity/wl128x-fw-5-mr.bin"
 #define WL128X_FW_NAME_SINGLE	"ti-connectivity/wl128x-fw-5-sr.bin"
diff --git a/net/compat-drivers/drivers/net/wireless/ti/wl12xx/wl12xx.h b/net/compat-drivers/drivers/net/wireless/ti/wl12xx/wl12xx.h
index 076c1d5..d455285 100644
--- a/net/compat-drivers/drivers/net/wireless/ti/wl12xx/wl12xx.h
+++ b/net/compat-drivers/drivers/net/wireless/ti/wl12xx/wl12xx.h
@@ -36,12 +36,12 @@
 #define WL127X_IFTYPE_SR_VER	3
 #define WL127X_MAJOR_SR_VER	10
 #define WL127X_SUBTYPE_SR_VER	WLCORE_FW_VER_IGNORE
-#define WL127X_MINOR_SR_VER	133
+#define WL127X_MINOR_SR_VER	115
 /* minimum multi-role FW version for wl127x */
 #define WL127X_IFTYPE_MR_VER	5
 #define WL127X_MAJOR_MR_VER	7
 #define WL127X_SUBTYPE_MR_VER	WLCORE_FW_VER_IGNORE
-#define WL127X_MINOR_MR_VER	42
+#define WL127X_MINOR_MR_VER	115
 
 /* FW chip version for wl128x */
 #define WL128X_CHIP_VER		7
