diff --git a/include/linux/mmc/host.h b/include/linux/mmc/host.h
index 1a8d714..66deb4bf 100644
--- a/include/linux/mmc/host.h
+++ b/include/linux/mmc/host.h
@@ -407,9 +407,8 @@ static inline void mmc_signal_sdio_irq(struct mmc_host *host)
 {
 	host->ops->enable_sdio_irq(host, 0);
 	host->sdio_irq_pending = true;
-#ifndef CONFIG_SEMC_MMC_SDIO_NO_IRQ
-	wake_up_process(host->sdio_irq_thread);
-#endif
+	if (host->sdio_irq_thread)
+		wake_up_process(host->sdio_irq_thread);
 }
 
 struct regulator;
