diff --git a/msm7x30.mk b/msm7x30.mk
index e6f0d6d..760fc14 100644
--- a/msm7x30.mk
+++ b/msm7x30.mk
@@ -119,7 +119,7 @@ PRODUCT_PROPERTY_OVERRIDES += \
     hwui.disable_vsync=true \
     debug.mdpcomp.logs=0 \
     BUILD_UTC_DATE=0 \
-    persist.sys.usb.config=mtp \
+    persist.sys.usb.config=mass_storage \
     debug.camcorder.disablemeta=1 \
     ro.debuggable=1 \
     persist.sys.strictmode.visual=0 \
