diff --git a/src/com/android/mms/data/WorkingMessage.java b/src/com/android/mms/data/WorkingMessage.java
index 250a63a..83ba3ef 100755
--- a/src/com/android/mms/data/WorkingMessage.java
+++ b/src/com/android/mms/data/WorkingMessage.java
@@ -1522,7 +1522,7 @@ public class WorkingMessage {
 
         Uri uri;
         try {
-            if (cursor.moveToFirst()) {
+            if (cursor != null && cursor.moveToFirst()) {
                 uri = ContentUris.withAppendedId(Mms.Draft.CONTENT_URI,
                         cursor.getLong(MMS_ID_INDEX));
                 String subject = MessageUtils.extractEncStrFromCursor(cursor, MMS_SUBJECT_INDEX,
@@ -1536,7 +1536,8 @@ public class WorkingMessage {
                 return uri;
             }
         } finally {
-            cursor.close();
+            if (cursor != null)
+                cursor.close();
         }
 
         return null;
diff --git a/src/com/android/mms/ui/ComposeMessageActivity.java b/src/com/android/mms/ui/ComposeMessageActivity.java
index 9758ee7..b3ceddb 100644
--- a/src/com/android/mms/ui/ComposeMessageActivity.java
+++ b/src/com/android/mms/ui/ComposeMessageActivity.java
@@ -4347,12 +4347,14 @@ public class ComposeMessageActivity extends Activity
                     int newSelectionPos = -1;
                     long targetMsgId = getIntent().getLongExtra("select_id", -1);
                     if (targetMsgId != -1) {
-                        cursor.moveToPosition(-1);
-                        while (cursor.moveToNext()) {
-                            long msgId = cursor.getLong(COLUMN_ID);
-                            if (msgId == targetMsgId) {
-                                newSelectionPos = cursor.getPosition();
-                                break;
+                        if (cursor != null) {
+                            cursor.moveToPosition(-1);
+                            while (cursor.moveToNext()) {
+                                long msgId = cursor.getLong(COLUMN_ID);
+                                if (msgId == targetMsgId) {
+                                    newSelectionPos = cursor.getPosition();
+                                    break;
+                                }
                             }
                         }
                     } else if (mSavedScrollPosition != -1) {
@@ -4408,7 +4410,7 @@ public class ComposeMessageActivity extends Activity
                     // mSentMessage is true).
                     // Show the recipients editor to give the user a chance to add
                     // more people before the conversation begins.
-                    if (cursor.getCount() == 0 && !isRecipientsEditorVisible() && !mSentMessage) {
+                    if (cursor != null && cursor.getCount() == 0 && !isRecipientsEditorVisible() && !mSentMessage) {
                         initRecipientsEditor();
                     }
 
@@ -4444,7 +4446,7 @@ public class ComposeMessageActivity extends Activity
                     if (cursor == null) {
                         return;
                     }
-                    if (tid > 0 && cursor.getCount() == 0) {
+                    if (tid > 0 && cursor != null && cursor.getCount() == 0) {
                         // We just deleted the last message and the thread will get deleted
                         // by a trigger in the database. Clear the threadId so next time we
                         // need the threadId a new thread will get created.
