diff --git a/prebuilt/init.semc.rc b/prebuilt/init.semc.rc
index d2ea04f..5fdd385 100644
--- a/prebuilt/init.semc.rc
+++ b/prebuilt/init.semc.rc
@@ -215,6 +226,7 @@ service hw_config /system/bin/sh /system/etc/hw_config.sh
 
 on property:dhcp.wlan0.result=ok
     exec /system/bin/iw phy phy0 wowlan enable any
+    write /sys/kernel/debug/ieee80211/phy0/wlcore/sleep_auth 1
 
 on property:init.svc.hostapd_bin=stopped
     exec /system/bin/iw dev mon.wlan0 del
