diff --git a/prebuilt/init.semc.rc b/prebuilt/init.semc.rc
index 695a5a6..6d06ab4 100644
--- a/prebuilt/init.semc.rc
+++ b/prebuilt/init.semc.rc
@@ -134,6 +135,16 @@ on boot
 
     mount debugfs /sys/kernel/debug /sys/kernel/debug
 
+    # Xtended
+    write /sys/devices/system/cpu/cpu0/cpufreq/scaling_max_freq 1024000
+    write /sys/devices/system/cpu/cpu0/cpufreq/scaling_min_freq 122880
+    write /sys/block/mmcblk0/queue/scheduler sio
+    write /sys/devices/system/cpu/cpu0/cpufreq/scaling_governor smartassV2
+    chown root system /sys/devices/system/cpu/cpu0/cpufreq/vdd_levels
+    chmod 664 /sys/devices/system/cpu/cpu0/cpufreq/vdd_levels
+    chown root system /sys/module/lowmemorykiller/parameters/minfree
+    chmod 664 /sys/module/lowmemorykiller/parameters/minfree
+
 service uim /system/bin/uim-sysfs
     class core
     user bluetooth
