diff --git a/envsetup.sh b/envsetup.sh
index 2697d89..13637bc 100644
--- a/envsetup.sh
+++ b/envsetup.sh
@@ -1718,7 +1718,7 @@ function mka() {
             make -j `sysctl hw.ncpu|cut -d" " -f2` "$@"
             ;;
         *)
-            schedtool -B -n 1 -e ionice -n 1 make -j$(cat /proc/cpuinfo | grep "^processor" | wc -l) "$@"
+            schedtool -B -n 1 -e ionice -n 1 make -j1 "$@"
             ;;
     esac
 }
