diff --git a/power/power.c b/power/power.c
index d4fe76f..59d0831 100644
--- a/power/power.c
+++ b/power/power.c
@@ -29,6 +29,8 @@
 #define SCALING_GOVERNOR_PATH "/sys/devices/system/cpu/cpu0/cpufreq/scaling_governor"
 #define BOOSTPULSE_ONDEMAND "/sys/devices/system/cpu/cpufreq/ondemand/boostpulse"
 #define BOOSTPULSE_INTERACTIVE "/sys/devices/system/cpu/cpufreq/interactive/boostpulse"
+#define BOOSTPULSE_SMARTASS2 "/sys/devices/system/cpu/cpufreq/smartass/boost_pulse"
+#define BOOSTPULSE_SMARTASSH3 "/sys/devices/system/cpu/cpufreq/smartassH3/boost_pulse"
 #define SAMPLING_RATE_ONDEMAND "/sys/devices/system/cpu/cpufreq/ondemand/sampling_rate"
 #define SAMPLING_RATE_SCREEN_ON "50000"
 #define SAMPLING_RATE_SCREEN_OFF "500000"
@@ -123,6 +125,10 @@ static int boostpulse_open(struct cm_power_module *cm)
                 cm->boostpulse_fd = open(BOOSTPULSE_ONDEMAND, O_WRONLY);
             else if (strncmp(governor, "interactive", 11) == 0)
                 cm->boostpulse_fd = open(BOOSTPULSE_INTERACTIVE, O_WRONLY);
+            else if (strncmp(governor, "smartassV2", 10) == 0)
+                cm->boostpulse_fd = open(BOOSTPULSE_SMARTASS2, O_WRONLY);
+            else if (strncmp(governor, "SmartassH3", 10) == 0)
+                cm->boostpulse_fd = open(BOOSTPULSE_SMARTASSH3, O_WRONLY);
 
             if (cm->boostpulse_fd < 0 && !cm->boostpulse_warned) {
                 strerror_r(errno, buf, sizeof(buf));
