diff --git a/adb_install.c b/adb_install.c
index 00c3cc0..1b8ac31 100644
--- a/adb_install.c
+++ b/adb_install.c
@@ -60,7 +60,7 @@ set_usb_driver(int enabled) {
 
 static void
 stop_adbd() {
-    property_set("ctl.stop", "adbd");
+    __system("killall adbd");
     set_usb_driver(0);
 }
 
@@ -72,7 +72,7 @@ maybe_restart_adbd() {
     if (len == 1 && value[0] == '1') {
         ui_print("Restarting adbd...\n");
         set_usb_driver(1);
-        property_set("ctl.start", "adbd");
+        __system("/sbin/adbd &");
     }
 }
 
@@ -129,7 +129,7 @@ apply_from_adb() {
 
     // kill the child
     kill(data.child, SIGTERM);
-    pthread_join(&sideload_thread, NULL);
+    pthread_join(sideload_thread, NULL);
     ui_clear_key_queue();
 
     struct stat st;
